### PR TITLE
[FEATURE] Search result struct

### PR DESCRIPTION
**Pull request by eseiler**
part of #1434

### CODE REVIEW

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 1]
```diff
@@ -123,38 +125,58 @@ inline auto search_single(index_t const & index, query_t & query, configuration_
 
     // TODO: filter hits and only do it when necessary (depending on error types)
 
+    constexpr bool output_cursor = cfg_t::template exists<search_cfg::output<detail::search_output_index_cursor>>();
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search.hpp]: I have started to make a global traits type with all necessary types and constants defined based on the configuration. This way you don't need to expose this cumbersome `template exists` syntax. I found this a little easier to read and to use throughout the entire code base

Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search.hpp]:
```suggestion
    constexpr bool requires_output_curser = cfg_t::template exists<search_cfg::output<detail::search_output_index_cursor>>();
```
?

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 2]
```diff
@@ -123,38 +125,58 @@ inline auto search_single(index_t const & index, query_t & query, configuration_
...
+                                        search_result_value_type<typename index_t::cursor_type, size_t, size_t>>;
+
+    std::vector<result_t> results;
+
     // output cursors or text_positions
-    if constexpr (cfg_t::template exists<search_cfg::output<detail::search_output_index_cursor>>())
+    if constexpr (output_cursor)
     {
-        return internal_hits;
+        for (auto && cursor : internal_hits)
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search.hpp]: so the hits are already generated? And here we copy them again? Or can the internal representation of internal hits already changed? Would that make sense?

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 3]
```diff
@@ -123,38 +125,58 @@ inline auto search_single(index_t const & index, query_t & query, configuration_
...
+        {
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search.hpp]: please remove the braces for single line bodies

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 4]
```diff
@@ -123,38 +125,58 @@ inline auto search_single(index_t const & index, query_t & query, configuration_
...
         {
-            for (auto const & cur : internal_hits)
+            // Store the results of locate (either `size_t` or `std::pair<size_t, size_t>`) for erasing duplicates.
+            decltype(internal_hits[0].locate()) hits;
+
+            for (auto && cursor : internal_hits)
             {
-                for (auto const & text_pos : cur.locate())
+                for (auto const & text_pos : cursor.locate())
+                {
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search.hpp]: remove braces

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 5]
```diff
@@ -123,38 +125,58 @@ inline auto search_single(index_t const & index, query_t & query, configuration_
...
         else
         {
-            for (auto const & cur : internal_hits)
+            // Store the results of locate (either `size_t` or `std::pair<size_t, size_t>`) for erasing duplicates.
+            decltype(internal_hits[0].locate()) hits;
+
+            for (auto && cursor : internal_hits)
             {
-                for (auto const & text_pos : cur.locate())
+                for (auto const & text_pos : cursor.locate())
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search.hpp]: position is an integral type? Then you can also capture it by value.

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 6]
```diff
@@ -123,38 +125,58 @@ inline auto search_single(index_t const & index, query_t & query, configuration_
...
+                {
                     hits.push_back(text_pos);
+                }
+
                 std::sort(hits.begin(), hits.end());
                 hits.erase(std::unique(hits.begin(), hits.end()), hits.end());
+
+                for (auto const & text_pos : hits)
+                {
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search.hpp]: remove braces

File: include/seqan3/search/algorithm/detail/search.hpp [hunk 7]
```diff
@@ -123,38 +125,58 @@ inline auto search_single(index_t const & index, query_t & query, configuration_
 
     // TODO: filter hits and only do it when necessary (depending on error types)
 
+    constexpr bool output_cursor = cfg_t::template exists<search_cfg::output<detail::search_output_index_cursor>>();
+
+    using result_t = std::conditional_t<output_cursor,
+                                        search_result_value_type<typename index_t::cursor_type>,
+                                        search_result_value_type<typename index_t::cursor_type, size_t, size_t>>;
+
+    std::vector<result_t> results;
```
Comment (rrahn) [inline include/seqan3/search/algorithm/detail/search.hpp]: Shouldn't this be `std::vector<search_result<search_result_value_t>>`?

File: include/seqan3/search/algorithm/search_result.hpp [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        query_id(_id), iterator(_it) {}
+
+    // Only one size_t that indicates sequence position. sequence id is then 0.
+    search_result_value_type(size_t _id, iterator_t _it, sequence_pos_t _pos) :
+    query_id(_id), iterator(_it), sequence_id(0), sequence_pos(_pos) {}
+
+    // Search returns a pair of sequence if and position.
+    search_result_value_type(size_t _id, iterator_t _it, std::pair<sequence_id_t, sequence_pos_t> _pair) :
+        query_id(_id), iterator(_it), sequence_id(_pair.first), sequence_pos(_pair.second) {}
+
```
Comment (rrahn) [inline include/seqan3/search/algorithm/search_result.hpp]: empty lines

File: include/seqan3/search/algorithm/search_result.hpp [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+          typename sequence_id_t = std::nullopt_t *,
+          typename sequence_pos_t = std::nullopt_t *>
+struct search_result_value_type
+{
+    size_t query_id{};
+    iterator_t iterator{};
+    sequence_id_t sequence_id{};
+    sequence_pos_t sequence_pos{};
+
+    search_result_value_type() = default;                                                //!< Defaulted
```
Comment (rrahn) [inline include/seqan3/search/algorithm/search_result.hpp]: missing dot at end of Comment

File: include/seqan3/search/algorithm/search_result.hpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+{
+
+template <typename search_result_traits>
+    requires detail::is_type_specialisation_of_v<search_result_traits, detail::search_result_value_type>
+class search_result
+{
+private:
+    search_result_traits data;
+
+    using iterator_t          = decltype(data.iterator);
```
Comment (rrahn) [inline include/seqan3/search/algorithm/search_result.hpp]: I would prefer not aligning the stuff. 
I mean if the documentation will be added this won't matter anyway so much

File: include/seqan3/search/algorithm/search_result.hpp [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        return data.sequence_id;
+    }
+
+    constexpr sequence_pos_t sequence_pos() const noexcept
+    {
+        static_assert(!std::is_same_v<sequence_pos_t, std::nullopt_t *>,
+                      "Identifier is not available but should.");
+        return data.sequence_pos;
+    }
+
```
Comment (rrahn) [inline include/seqan3/search/algorithm/search_result.hpp]: empty line

File: include/seqan3/search/algorithm/detail/search.hpp (hunk 2 above)
Comment (eseiler) [reply include/seqan3/search/algorithm/detail/search.hpp]: We have the cursors and need to move them into the `search_result_value_type`.... doesn't this move? `internal_hits` will go out of scope after this and won't change anymore.

File: include/seqan3/search/algorithm/detail/search.hpp (hunk 5 above)
Comment (eseiler) [reply include/seqan3/search/algorithm/detail/search.hpp]: It's `size_t` or `std::pair<size_t, size_t>`

File: include/seqan3/search/algorithm/detail/search.hpp (hunk 7 above)
Comment (eseiler) [reply include/seqan3/search/algorithm/detail/search.hpp]: I'd like to, but we need to enumerate the queries in the caller `search_all()` and hence need to modify the `search_result_value_type`

File: include/seqan3/search/algorithm/search_result.hpp [hunk 12]
```diff
@@ -0,0 +1,104 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
```
Comment (smehringer) [inline include/seqan3/search/algorithm/search_result.hpp]: don't forget to update the year

**Comment by rrahn** [comment]
I believe this is superseded by #1706? Please, feel free to reopen if this is not the case.

**Pull request closed without merging**